lib/purescript-cst/src/Language/PureScript/CST/Layout.hs

```diff
@@ -17,7 +24,144 @@
 -- such an algorithm. Unquoted properties for layout keywords introduce a domino
 -- effect of complexity since we have to mask and unmask any usage of . (also in
 -- foralls!) or labels in record literals.
-
+--
+-- ## Detailed Summary
+--
+-- ### The Problem
+--
+-- The parser itself is unaware of indentation or other such layout concerns.
+-- Rather than dealing with it explicitly, the parser and its
+-- grammar rules are only aware of normal tokens (e.g. @TokLowerName@) and
+-- three special zero-width tokens, @TokLayoutStart@, @TokLayoutSep@,
+-- and @TokLayoutEnd@. This is convenient because the actual grammar
+-- can be specified apart from the indentation rules and other such
+-- layout concerns.
+--
+-- For a simple example, the parser parses all three examples of the code below
+-- using the exact same grammar rules for the @let@ keyword despite
+-- each example using different indentations levels:
+--
+-- @
+-- -- Example 1
+-- let foo = 5 in foo
+--
+-- -- Example 2
+-- let
+--   bar = 5
+-- in bar
+--
+-- -- Example 3
+-- let        baz
+--                  =
+--              5 in baz
+-- @
+--
+-- Each block of code might appear to the parser as a stream of the
+-- following source tokens where the @\{@ sequence represents
+-- @TokLayoutStart@ and the @\}@ sequence represents @TokLayoutEnd@:
+-- - @\{let\{foo\}=\{5\}\}@
+-- - @\{let\{bar\}=\{5\}\}@
+-- - @\{let\{baz\}=\{5\}\}@
```

Comment:
This is likely wrong as to how many of these tokens are inserted and where they are inserted. My goal here was to simply show that they are being inserted and that whitespace, comments, and lines are irrelevant.

Reply:
It would be
```
let \{foo = 5\}
```
I think you should include multiple definitions, eg:
```
let \{foo = 5\;bar = 42\} in ...
```

Reply:
Done.